programs/metadata/src/lib.rs

```diff
@@ -0,0 +1,195 @@
+        metadata.creation_slot = Clock::get()?.slot;
+        metadata.last_updated_slot = Clock::get()?.slot;
+        metadata.items = Vec::new();
+        Ok(())
+    }
+
+    pub fn increase_metadata_account_size(
+        _ctx: Context<IncreaseMetadataAccountSize>,
+    ) -> Result<()> {
+        Ok(())
+    }
+
+    pub fn set_delegate(ctx: Context<SetDelegate>) -> Result<()> {
+        let metadata = &mut ctx.accounts.metadata;
+        metadata.delegate = ctx.accounts.new_delegate.key();
+        metadata.last_updated_slot = Clock::get()?.slot;
+        Ok(())
+    }
+
+    // Instruction to add a new MetadataItem
+    pub fn create_metadata_item(
+        ctx: Context<ModifyMetadata>,
+        key: String,
+        value: Vec<u8>,
+    ) -> Result<()> {
+        let metadata = &mut ctx.accounts.metadata;
+        require!(
+            metadata.items.iter().all(|item| item.key != key),
+            ErrorCode::DuplicateKey
+        );
+        let item = MetadataItem {
+            // update_authority: metadata.delegate,
+            last_updated_slot: Clock::get()?.slot,
+            key,
+            value,
+        };
+        metadata.items.push(item);
+        metadata.last_updated_slot = Clock::get()?.slot;
+        Ok(())
+    }
+
+    // Instruction to delete a MetadataItem
+    pub fn delete_metadata_item(ctx: Context<ModifyMetadata>, key: String) -> Result<()> {
+        let metadata = &mut ctx.accounts.metadata;
+        let current_slot = Clock::get()?.slot;
+        if let Some(item) = metadata.items.iter_mut().find(|item| item.key == key) {
+            require_gt!(
+                current_slot,
+                item.last_updated_slot,
+                ErrorCode::InvalidOperationInCurrentSlot
+            );
+            metadata.items.retain(|item| item.key != key);
+            metadata.last_updated_slot = Clock::get()?.slot;
+        } else {
+            msg!("Key: {}", key);
+            return Err(error!(ErrorCode::KeyNotFound));
+        }
+        Ok(())
+    }
+
+    // Instruction to overwrite data on an existing MetadataItem
+    pub fn write_metadata_item(
+        ctx: Context<ModifyMetadata>,
+        key: String,
+        new_value: Vec<u8>,
+    ) -> Result<()> {
+        let metadata = &mut ctx.accounts.metadata;
+        let current_slot = Clock::get()?.slot;
+        if let Some(item) = metadata.items.iter_mut().find(|item| item.key == key) {
+            require_gt!(
+                current_slot,
+                item.last_updated_slot,
+                ErrorCode::InvalidOperationInCurrentSlot
+            );
+            item.value = new_value;
+            item.last_updated_slot = current_slot;
+            metadata.last_updated_slot = current_slot;
+        } else {
+            msg!("Key: {}", key);
+            return Err(error!(ErrorCode::KeyNotFound));
+        }
+        Ok(())
+    }
+
+    // Instruction to append data to an existing MetadataItem
+    pub fn append_metadata_item(
```

Comment:
what if we took the buffer approach, so instead of writing you write into an offset? so no more append

Reply:
Is there a specific issue with append? Using vecs, no zero copy, is much easier than having the UI or the code here figure out where the next 0 slot was. Plus we'd have start allocating specific array length to all of the metadata items which are all a bit different.